sql-plugin/src/main/scala/com/nvidia/spark/rapids/PrioritySemaphore.scala

```diff
@@ -58,10 +58,21 @@ class PrioritySemaphore[T](val maxPermits: Int)(implicit ordering: Ordering[T])
     lock.lock()
     try {
       val condition = lock.newCondition()
+      var queued = false
       while (!canAcquire(numPermits)) {
-        waitingQueue.enqueue(ThreadInfo(priority, condition))
+        if (!queued) {
+          waitingQueue.enqueue(ThreadInfo(priority, condition, numPermits))
+          queued = true
+        }
         condition.await()
       }
+      if (queued) {
+        if (waitingQueue.isEmpty || waitingQueue.head.condition != condition) {
+          throw new IllegalStateException("acquire thought it queued but acquired semaphore "
+            + "without being at the head of the queue")
+        }
+        waitingQueue.dequeue()
```

Comment:
Only dequeue when we're really done (out of loop/semaphore acquired)

Reply:
I don't think we want the receiving thread to remove itself from the queue. It opens up the possibility of a race condition in release where two threads "release" their slots, but then they both wake up the same thread. This would result in us running with less parallelism than we are configured to run with. It would not be a deadlock, but it would also not be ideal.